pkg/bpf/collection.go

```diff
@@ -0,0 +1,251 @@
+//
+// It parses the ELF section name of each ProgramSpec to extract any map/slot
+// mappings for prog arrays used as tail call maps. The spec's programs are then
+// inserted into the appropriate map and slot.
+//
+// TODO(timo): Remove when bpf_elf_map map definitions are no longer used after
+// moving away from iproute2+libbpf.
+func iproute2Compat(spec *ebpf.CollectionSpec) error {
+	// Parse legacy iproute2 u32 id and pinning fields.
+	maps := make(map[uint32]*ebpf.MapSpec)
+	for _, m := range spec.Maps {
+		if m.Extra != nil && m.Extra.Len() > 0 {
+			tail := struct {
+				ID      uint32
+				Pinning uint32
+				_       uint64 // inner_id + inner_idx
+			}{}
+			if err := binary.Read(m.Extra, spec.ByteOrder, &tail); err != nil {
+				return fmt.Errorf("reading iproute2 map definition: %w", err)
+			}
+
+			if tail.Pinning > 0 {
+				m.Pinning = 1 // LIBBPF_PIN_BY_NAME
+			}
+
+			// Index maps by their iproute2 .id if any, so X/Y ELF section names can
+			// be matched against them.
+			if tail.ID != 0 {
+				if m2 := maps[tail.ID]; m2 != nil {
+					return fmt.Errorf("maps %s and %s have duplicate iproute2 map ID %d", m.Name, m2.Name, tail.ID)
+				}
+				maps[tail.ID] = m
+			}
+		}
+	}
+
+	for n, p := range spec.Programs {
+		// Parse the program's section name to determine which prog array and slot it
+		// needs to be inserted into. For example, a section name of '2/14' means to
+		// insert into the map with the .id field of 2 at index 14.
+		// Uses %v to automatically detect slot's mathematical base, since they can
+		// appear either in dec or hex, e.g. 1/0x0515.
+		var id, slot uint32
+		if _, err := fmt.Sscanf(p.SectionName, "%d/%v", &id, &slot); err == nil {
+			// Assign the prog name and slot to the map with the iproute2 .id obtained
+			// from the program's section name. The lib will load the ProgramSpecs
+			// and insert the corresponding Programs into the prog array at load time.
+			m := maps[id]
+			if m == nil {
+				return fmt.Errorf("no map with iproute2 map .id %d", id)
+			}
+			m.Contents = append(maps[id].Contents, ebpf.MapKV{Key: slot, Value: n})
+		}
+	}
+
+	return nil
+}
+
+// LoadCollection loads the given spec into the kernel with the specified opts.
+//
+// Any maps marked as pinned in the spec are automatically loaded from the path
+// given in opts.Maps.PinPath and will be used instead of creating new ones.
+// MapSpecs that differ (type/key/value/max/flags) from their pinned versions
+// will result in an ebpf.ErrMapIncompatible here and the map must be removed
+// before loading the CollectionSpec.
+func LoadCollection(spec *ebpf.CollectionSpec, opts ebpf.CollectionOptions) (*ebpf.Collection, error) {
+	if spec == nil {
+		return nil, errors.New("can't load nil CollectionSpec")
+	}
+
+	// By default, allocate a 1MiB verifier log buffer if first load attempt
+	// fails. This was adjusted around Cilium 1.11 for fitting bpf_lxc insn
+	// limit messages.
+	if opts.Programs.LogSize == 0 {
```

Comment:
^ IIRC iproute2 includes all of the above logic, so to import the equivalent behaviour here we'll need to write the loop & retry logic ourselves.

Reply:
The library already incorporates this functionality, see https://github.com/cilium/ebpf/blob/629483784baa95e91c992134241439525cad0dac/prog.go#L274-L304.

By default, prog_load is run with log level 0, so there will be no verifier output. If that fails, it's retried with the verifier log enabled, and the buffer set to the default size or the LogSize specified in opts.

If the log level is non-zero and a buffer is provided, log full means -ENOSPC, but that's only when the program fails verification.

The problem with retrying with bigger and bigger buffers is that we can't really estimate how big the log will be. Given a default starting buffer of 65k, where do we go from there? Do we double it? 1MiB is a few retries off from there.. How often do we retry and when do we give up? I don't believe retrying with bigger logs is a good default strategy, at least not with bpf_lxc spending ~600ms in the verifier on my machine.

Instead, the user should provide a reasonable size. As mentioned in the comment, largest log I've seen by far is from bpf_lxc and was a bit under a MiB. Verifier errors should be comparatively rare outside of development, so if ENOSPC is encountered, it would most likely be caught before going to production. At that point, this value can be bumped. Rinse, repeat.

Reply:
That makes total sense, thanks for the explanation.

Reply:
btw minor nit here, the code specifies 1<<24 (16MiB) rather than 1<<20 (1MiB).

Reply:
Good catch. 1<<24 is 1000000 in hex, not decimal. :sweat: